glibc: 2.33-108 -> 2.34-115

Ma27 opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Announcement: https://sourceware.org/pipermail/libc-alpha/2021-August/129718.html

Hydra jobset: https://hydra.nixos.org/jobset/nixpkgs/glibc-2.34

__Note__ for reviewers: this update is slightly more nasty than the previous ones, especially the following commits should be reviewed closely:
* https://github.com/NixOS/nixpkgs/pull/133431/commits/97acaf6d65e869abf7dc9e6825aef4c0e7b98432 (`glibc: 2.33-108 -> 2.34-115`)
* https://github.com/NixOS/nixpkgs/pull/133431/commits/da905d4cf918e28f924bab99c6f2a4012d55410d (`nixos/stage-1: fix modprobe in initial ramdisk on systems w/glibc-2.34`)
* https://github.com/NixOS/nixpkgs/pull/133431/commits/748faeb623035c1360eeb6dc4b05abf5defd721c (`glibc: don't hide symbol __nss_files_fopen`

__Todos:__

* [x] Await current jobset (https://hydra.nixos.org/eval/1741006)
* [x] @Ma27 will try to rebuild their workstation with this PR to check if anything notable broke
* [x] Get reviews :) 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Ma27 committed:
glibc: 2.33-108 -> 2.34-115

Announcement: https://sourceware.org/pipermail/libc-alpha/2021-August/129718.html

Ma27 committed:
stdenv-bootstrap: force using new libc from stage2

This fix is needed to work around linker-errors such as

    undefined reference to `__libc_csu_fini'

which I got in almost every derivation which is part of stage2. The
reasoning behind this is that the startup-code was simplified[1] and
thus `__libc_csu_fini` doesn't exist anymore.

A workable solution is to use a newer libc which properly links in
stage3. And actually this seems expected given the rationale for stage3:

    # Construct a third stdenv identical to the 2nd, except that this
    # one uses the rebuilt Glibc from stage2.  It still uses the recent
    # binutils and rest of the bootstrap tools, including GCC.

So this patch basically overrides the libraries inside `gcc-unwrapped` -
which is basically the bootstrap tools and thus also contains the libc
used in stage3 - with the shared objects from the freshly built libc
from stage2.

[1] https://sourceware.org/pipermail/libc-alpha/2021-March/123079.html

Ma27 committed:
glibc: symlink `libpthread.so -> libpthread.so.0` (same for `-lrt`) for backwards compatibility

Linking via `-lpthread` (or `-pthread`) is not needed anymore since
`glibc-2.34` since all the functionality is part of `libc.so.6` and
`libpthread.so.6` only exists for backwards-compatibility.

However, e.g. `gcc` (`libgomp` to be precise) expects a `libpthread.so`
to link against, otherwise the configure script will fail. As already
stated in the glibc release-notes itself, it is to expect that a lot
more applications will have issues with this, so I decided to re-add
`libpthread.so` as well.

For `librt.so.1`, the same thing is needed to make sure that Perl still
compiles:

    /nix/store/d6y5r7m93x14bmgn2p75fannz39jz66f-binutils-2.35.1/bin/ld: cannot find -lrt
    collect2: error: ld returned 1 exit status
    make[1]: *** [Makefile:490: ../../lib/auto/Time/HiRes/HiRes.so] Error 1
    make[1]: Leaving directory '/build/perl-5.34.0/dist/Time-HiRes'

Ma27 committed:
findutils: fix build w/glibc-2.34

This "-D__nonnull\\(params\\)=" leads to a compilation failure in e.g.
the configure phase:

    configure:21131: gcc -c -D__nonnull\(params\)=  conftest.c >&5
    <command-line>: warning: ISO C99 requires whitespace after the macro name
    <command-line>: error: stray '\' in program
    <command-line>: error: expected ',' or ';' before '(' token
    <command-line>: error: stray '\' in program

According to the commit this isn't even needed on Linux.

I confirmed that this is an (expectable) glibc-2.34 thing by checking
that

* the issue doesn't occur with gcc 10/11 on a recent glibc-2.33 staging.
* the issue DOES occur in a docker container with Fedora rawhide (which
  has glibc 2.34 and gcc 11).

Ma27 commented:
@vcunat @lovesegfault I can build on this branch `pkgs.hello` now, but some of the changes are slightly messy. So before I continue here, I'd have two requests:

* It would be nice to get some feedback to the current changes I made (I tried to document the reasons in the corresponding commit messages)
* It would help if https://hydra.nixos.org/jobset/nixpkgs/glibc-2.33 could be reconfigured to be re-used for glibc-2.34.

Thanks! :)

SuperSandro2000 commented:
Does this require the binutils update?

Ma27 commented:
Doesn't seem as if that's the case currently :)

vcunat commented:
I'm not sure if it's good to test them separately and then merge without testing those two updates together.

vcunat commented:
Quick look: the bootstrapping change is OK-ish, I think.  The rest seems good.

I'm not sure why diffutils update was included.  Nit: `2.33-master.patch.gz` should be deleted.

Ma27 commented:
> I'm not sure if it's good to test them separately and then merge without testing those two updates together.

I haven't taken a close look at bintuils, but considering that I even had to do some workarounds to get gcc building (yay, compatibility), I guess this will take a while (no idea if it'll land in 21.11), so it's rather likely that this will be here for a while after a `git pull upstream staging --rebase` for a while :) 

> Quick look: the bootstrapping change is OK-ish, I think. The rest seems good.

OK, great! :) 

> I'm not sure why diffutils update was included

With this change, the package compiled again (but good point, should perhaps add a note to the commit message).

>  Nit: 2.33-master.patch.gz should be deleted

Oh, good point :)

Ma27 committed:
glibc: also create backwards-compat symlinks for libdl and libutil

This fixes at least the openssl build.

Ma27 committed:
boost1{69,70,72}: fix build w/glibc-2.34

To quote the release-notes[1]:

> When _DYNAMIC_STACK_SIZE_SOURCE or _GNU_SOURCE are defined,
> PTHREAD_STACK_MIN is no longer constant and is redefined to
> sysconf(_SC_THREAD_STACK_MIN).  This supports dynamic sized register
> sets for modern architectural features like Arm SVE.

This basically means that if the above applies, `#if PTHREAD_STACK_MIN > 0`
won't compile anymore because `PTHREAD_STACK_MIN` isn't a hard-coded
number, but `__sysconf (__SC_THREAD_STACK_MIN_VALUE)`[2].

The issue (for 1.69, 1.70, 1.72 - the other versions seem OK) was
reported upstream, but only for Solaris[3], however the corresponding
patches[4] seem to work as well for us.

Failing Hydra build: https://hydra.nixos.org/build/150926294

[1] https://sourceware.org/pipermail/libc-alpha/2021-August/129718.html
[2] See `${pkgs.glibc.dev}/include/bits/pthread_stack_min-dynamic.h`
[3] https://github.com/boostorg/thread/issues/283
[4] https://github.com/conan-io/conan-center-index/pull/361

Ma27 committed:
glibc: revert `/bin/bash` usage

Rather than `$BASH`, `glibc` now hardcodes `/bin/bash` as
interpreter[1] in several scripts (including `ldd`).

This is a problem because relevant programs such as `ldd(1)` won't work
properly without this since we set `BASH` to `/bin/sh` to avoid
dependencies to the bootstrap tools for `bash` as runtime-dependency (since
NixOS doesn't have `/bin/bash`).

Considering that this was only done as an "improvement" to their
build-system and not because they wanted to use some bashisms here (the
variable was always called `BASH` and we still used `/bin/sh` anyways),
I'd consider this to be relatively safe.

[1] 5188a9d0265cc6f7235a8af1d31ab02e4a24853d

vcunat commented:
I cancelled these builds for now, as too much work is piling up.  (also PR #135611)

Ma27 commented:
@vcunat Hydra's queue status currently says `Queue (117/125309)`. I have to admit that I don't really know the average and I didn't follow the current state of what's going on (with openssl I guess?) very closely (exam phase ^^), but I assume I should wait a few days before starting a new eval? :)

vcunat commented:
Yes, most of 21.05 jobs still don't have binaries with the new openssl.  And then there's lots of stuff that was cancelled and will also want to continue after this settles, e.g. the current `staging-next`.

r-burns commented:
fetchpatchable patch for catch 1.x: https://src.fedoraproject.org/rpms/catch1/blob/059104ba87494c0b5ebe16844ec190f253e51cac/f/catch1-sigstksz.patch

Ma27 committed:
libressl: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/151984996

vcunat commented:
You want to cancel the outdated builds, right?

I also restricted the platforms to just x86-linux, so that the bash-upgrade jobset can pre-test on aarch64-linux... assuming that most of the issues would be the same on both platforms (Hydra's capacity is limited).

Ma27 commented:
> You want (me) to cancel the outdated builds, right?

Oof, sorry forgot about that! Yes please! :)

Ma27 commented:
@vcunat would you mind hitting "Restart all failed jobs" on https://hydra.nixos.org/eval/1702817? Seems as if most of the builds are broken because of https://hydra.nixos.org/build/152422523 which seems to be fairly transient (https://hydra.nixos.org/log/vizgh9p9qc3v9f4c17f10smqb7zg90sg-libssh2-1.10.0.tar.gz.drv).

vcunat commented:
That's not really transient: https://github.com/NixOS/nix/issues/5089#issuecomment-915972254

vcunat commented:
Worked around and restarted.  (but x86_64-linux is relatively busy)

Ma27 committed:
fuse: fix build w/glibc-2.34

`closefrom` is now a part of `glibc-2.34`[1]:

> * The function closefrom has been added.  It closes all file descriptors
>   greater than or equal to a given integer.  This function is a GNU extension,
>   although it is also present in other systems.

Failing Hydra build: https://hydra.nixos.org/build/152456339

[1] https://sourceware.org/pipermail/libc-alpha/2021-August/129718.html

Ma27 committed:
catch: fix build w/glibc-2.34

The issue was that `SIGSTKSZ` isn't an actual const anymore and thus
can't be used to define sizes of static variables such as

    static char foo[SIGSTKSZ];

since this results in compiler errors such as

    error: array bound is not an integer constant before ']' token

Fedora worked around this by hard-coding the value in `catch`. Since
this is mainly a testing-framework and there's no other fix for v1 - we
should eventually remove it entirely in favor of v2 anyways - I guess
this is a good-enough fix.

Failing Hydra build: https://hydra.nixos.org/build/152455108

Ma27 committed:
autofs: fix build w/glibc-2.34

This build was also broken by a libc const that isn't a number anymore
and thus can't be used at places where a constant value is needed:

    automount.c:86:37: error: initializer element is not constant

Failing Hydra build: https://hydra.nixos.org/build/153253104

Ma27 committed:
qt515.qtwebengine: fix build w/glibc-2.34

Derived from Fedora 35: https://src.fedoraproject.org/rpms/qt5-qtwebengine/c/d122c011631137b79455850c363676c655cf9e09?branch=rawhide
Failing Hydra build: https://hydra.nixos.org/build/153232768

Ma27 committed:
qt514.qtwebengine: fix build w/glibc-2.34

Same problem here as with QT 5.15.

Failing Hydra build: https://hydra.nixos.org/build/153224604

Ma27 committed:
rcs: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/153235059

Ma27 commented:
At this point I should actually give a shout-out to the Fedora folks who already fixed a lot of glibc-2.34 issues. Quite a number of userland packages can be fixed now by applying their patches :heart:

vcunat commented:
Yes, Fedora and Arch tend to be quite good sources for bleeding-edge patches.

Ma27 committed:
emacs: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/153252802

Ma27 committed:
gdb: fix build w/glibc-2.34

This is a problem that seems to be related to the most recent `gdb`
update in staging from 10.2 to 11.1[1] where `gdb` started to fail
during build with the following message:

    checking for stdlib.h... In file included from /nix/store/vf96x4h90fm7bwf5zvfx8zb82fm1p21j-glibc-2.34-5-dev/include/signal.h:328,
                     from ../../gnulib/import/signal.h:52,
                     from targ-map.c:7:
    targ-map.c:412:17: error: initializer element is not constant
      412 |   { "SIGSTKSZ", SIGSTKSZ, TARGET_SIGSTKSZ },
          |                 ^~~~~~~~
    targ-map.c:412:17: note: (near initialization for 'cb_init_signal_map[18].host_val')

Since I couldn't find any patches in the upstream repo or for other
repos - according to Repology we seem to be the only distro trying to
ship `gdb-11` with `glibc-2.34` - so I found the culprit while bisecting
`gdb` which seems to be commit `a0e674c1`[2].

It seems as if the entire `sim/`-subtree is now built by default if
`--enable-targets=all` is set (which we do for cross debugging). However
it also generates a file called `targ-map.c` referencing `SIGSTKSZ`
assuming that it's const, although this is not the case anymore with
`glibc-2.34`[3].

Since I don't really understand, what precisely is going on in there and
there are no patches available I decided to switch back to the 10.2
behavior here and disable the feature by specifying `--disable-sim` as
configure flag.

Failing Hydra build: https://hydra.nixos.org/build/153893135

[1] 43b96f66ef7a69602b74f9bced64573c11a1df05
[2] https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=a0e674c1ce2c877426f8a861c5294c535c5d49e6
[3] see https://sourceware.org/pipermail/libc-alpha/2021-August/129718.html

Ma27 commented:
Cleaned up the history a bit to keep it bisectable to a certain degree and fixed the gdb 11 regression that was introduced with the latest staging rebase and used that chance to bump glibc to patch-level 9.

@vcunat would you mind stopping all pending builds in https://hydra.nixos.org/eval/1706506 and start a new eval? :)

vcunat commented:
Done.

Ma27 committed:
postfix: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/154082134

Ma27 committed:
ocaml: fix build w/glibc-2.34

ChangeLog: https://hydra.nixos.org/build/154122673

Ma27 committed:
openjdk11: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/154132196

Ma27 committed:
texinfo: fix build w/glibc-2.34

Ma27 committed:
ocaml 4.10/4.11: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155189331
Applied a smaller patch from Fedora's fork[1] as it also applies on
older versions.

That said, these versions seem unmaintained anyways, so we should
probably drop them entirely.

[1] https://pagure.io/fedora-ocaml/c/dfb5e954a04f59b0456cc4c0ddf3acaf22e0ff07?branch=fedora-35-4.12.0

Ma27 committed:
spdlog: 1.8.5 -> 1.9.2, fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155163350

Ma27 committed:
seasocks: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155145624

Ma27 committed:
leatherman: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155215198

Ma27 committed:
eternal-terminal: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/154077447

Ma27 committed:
trenchbroom: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155141010

Artturin commented:
i checked out this pr and applied #141201 as a patch and tinycc compiled

Ma27 committed:
trafficserver: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155205200

Ma27 committed:
pdfslicer: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155183850

Ma27 committed:
openmw: mark as broken

Ma27 committed:
ntp: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155170191

Patch derived from linuxfromscratch/openembedded.org[1][2].

[1] https://www.linuxfromscratch.org/blfs/view/svn/basicnet/ntp.html
[2] https://patchwork.openembedded.org/patch/180019/

Ma27 committed:
breakpad: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155160812

Fix derived from Fedora's Chromium fix[1].

[1] https://src.fedoraproject.org/rpms/chromium/blob/c3fea076996d62bff1404d06de5ef3adf93426db/f/chromium-92.0.4515.107-rawhide-gcc-std-max-fix.patch

Ma27 committed:
arangodb*: fix build w/glibc-2.34

Failing Hydra builds:
* https://hydra.nixos.org/build/155187495 (3.3)
* https://hydra.nixos.org/build/155161270 (3.4)
* https://hydra.nixos.org/build/155154245 (3.5)

Ma27 committed:
apitrace: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155150225

Ma27 committed:
aspcud: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155189149

Ma27 committed:
conky: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/155171697

Ma27 committed:
polyml*: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/156205993

Ma27 committed:
nixos/stage-1: fix `modprobe` in initial ramdisk on systems w/glibc-2.34

This effectively fixes the majority of all VM tests which were broken
because `/dev/vda` (or any other block device) wasn't mountable:

      machine # mounting /dev/vda on /...
      machine # mount: mounting /dev/vda on /mnt-root/ failed: No such device[    2.820976] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
      machine # [    2.821757] CPU: 0 PID: 1 Comm: init Not tainted 5.10.72 #1-NixOS
      machine # [    2.821757] Hardware name: QEMU Standard PC (i440FX + PIIX, 1996), BIOS rel-1.14.0-0-g155821a1990b-prebuilt.qemu.org 04/01/2014
      machine # [    2.821757] Call Trace:
      machine # [    2.821757]  dump_stack+0x6b/0x83
      machine # [    2.821757]  panic+0x101/0x2c8
      machine # [    2.821757]  do_exit.cold+0x14/0xb3
      machine # [    2.821757]  do_group_exit+0x33/0xa0
      machine # [    2.821757]  __x64_sys_exit_group+0x14/0x20
      machine # [    2.821757]  do_syscall_64+0x33/0x40
      machine # [    2.821757]  entry_SYSCALL_64_after_hwframe+0x44/0xa9
      machine # [    2.821757] RIP: 0033:0x7f67ec2800f6
      machine # [    2.821757] Code: 00 4c 8b 0d 2c 5d 11 00 eb 19 66 2e 0f 1f 84 00 00 00 00 00 89 d7 89 f0 0f 05 48 3d 00 f0 ff ff 77 22 f4 89 d7 44 89 c0 0f 05 <48> 3d 00 f0 ff ff 76 e2 f7 d8 64 41 89 01 eb da 66 2e 0f 1f 84 00
      machine # [    2.821757] RSP: 002b:00007fff8f5a71d8 EFLAGS: 00000202 ORIG_RAX: 00000000000000e7
      machine # [    2.821757] RAX: ffffffffffffffda RBX: 0000000000699704 RCX: 00007f67ec2800f6
      machine # [    2.821757] RDX: 0000000000000001 RSI: 000000000000003c RDI: 0000000000000001
      machine # [    2.821757] RBP: 0000000000000004 R08: 00000000000000e7 R09: ffffffffffffff80
      machine # [    2.821757] R10: 00007f67ec33f3e0 R11: 0000000000000202 R12: 000000000000000b
      machine # [    2.821757] R13: 00007fff8f5a75a8 R14: 0000000000000000 R15: 00000000004fc198
      machine # [    2.821757] Kernel Offset: 0x31e00000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
      machine # [    2.821757] Rebooting in 1 seconds..

This happened because the kernel failed to load modules such as `ext4`
from `boot.initrd.availableKernelModules`[1] on e.g. a `mount(2)` syscall.

The problem is that `kmod` isn't linked against `libpthread.so.0`
anymore because it got merged into `libc.so.6` (however, the .so still
exists), but still needs it:

      machine # newfstatat(AT_FDCWD, "/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-glibc-2.34-36/lib/x86_64", 0x7ffd951114c0, 0) = -1 ENOENT (No such file or directory)
      machine # openat(AT_FDCWD, "/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-glibc-2.34-36/lib/x86_64/libpthread.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
      machine # newfstatat(AT_FDCWD, "/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-glibc-2.34-36/lib/x86_64", 0x7ffd951114c0, 0) = -1 ENOENT (No such file or directory)
      machine # openat(AT_FDCWD, "/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-glibc-2.34-36/lib/libpthread.so.0", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
      machine # newfstatat(AT_FDCWD, "/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-glibc-2.34-36/lib", 0x7ffd951114c0, 0) = -1 ENOENT (No such file or directory)
      machine # openat(AT_FDCWD, "/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-glibc-2.34-36/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
      machine # writev(2, [{iov_base="/nix/store/kdc9n48ksdc1a8y8w512w"..., iov_len=69}, {iov_base=": ", iov_len=2}, {iov_base="error while loading shared libra"..., iov_len=36}, {iov_base=": ", iov_len=2}, {iov_base="libpthread.so.0", iov_len=15}, {iov_base=": ", iov_len=2}, {iov_base="cy
      machine # ) = 184
      machine # exit_group(127)                         = ?
      machine # +++ exited with 127 +++
      machine # mount: mounting /dev/vda on /mnt-root/ failed: No such device
      machine # [   19.167180] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
      machine # [   19.167711] CPU: 0 PID: 1 Comm: init Not tainted 5.10.72 #1-NixOS

This is not a problem

* inside stage-1 because `LD_LIBRARY_PATH` points to `$out/lib` of
  extra-utils where `libpthread.so.6` also exists.
* on a running system because `${pkgs.glibc}/lib` is part of kmod's
  rpath.

However this is a problem inside the kernel which calls `modprobe` (in
our case `kmod`) to load modules and doesn't know about
`LD_LIBRARY_PATH`. Also, the rpath-reference was nuked.

To work around this, the kernel's `modprobe`
(i.e. `/proc/sys/kernel/modprobe`) now points to a wrapper which
explicitly declares `LD_LIBRARY_PATH`. We can't use `makeWrapper` here
because `modprobe` itself must not be renamed. Otherwise, `kmod` (which
is the link-target of `modprobe`) won't work because it expects
`argv[0] == "modprobe"` to perform modprobe's tasks.

[1] https://nixos.org/manual/nixos/stable/options.html#opt-boot.initrd.availableKernelModules

Ma27 committed:
libowfat: mark as broken

Failing Hydra build: https://hydra.nixos.org/build/156248131

This doesn't build with `glibc-2.34`, however there's already an
upstream fix and can be marked as unbroken as soon as it's available[1].

I tried to apply the patch from a Git mirror[2], however this doesn't
apply properly on `0.32` and I decided to wait for a release - such as
Gentoo does it - instead.

[1] https://bugs.gentoo.org/806505
[2] https://git.lighttpd.net/mirrors/libowfat/commit/a10f89d5f3fbe58e8233cc755b599d1d07c63e44

Ma27 committed:
tiscamera: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/156249703

Ma27 committed:
ursadb: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/156279394

Ma27 committed:
libfive: mark as broken

Failing Hydra build: https://hydra.nixos.org/build/156227205

Vendored `catch2` is too old and thus cannot be patched.

Ma27 committed:
libspf2: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/156229214

Ma27 committed:
nethack-qt: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/156236800

Ma27 commented:
On the last Hydra eval there was a networkmanager regression that appears to stem from staging. Just rebased locally and it seems gone. Currently checking if stdenv builds fine, then I'd push and start a new eval.

I think most of the breaking things should be fixed then, so this is mergable in principle (but obviously has to wait until the feature-freeze period is finished).

fabianhjr commented:
Hey, slight bump to this. Freeze and branchoff of `21.11` already happened and hydra is mostly green but from a month ago ( https://hydra.nixos.org/jobset/nixpkgs/glibc-2.34 )

Ma27 commented:
Rebased and started a new eval :)

Ma27 commented:
OK missed that elfutils has the patch already in the new release, sorry!

@vcunat would you mind cancelling all pending jobs and schedule a new eval? :)

Ma27 commented:
I guess I found a major blocker: `__nss_files_fopen` cannot be wrapped anymore as the way it's loaded is changed internally which affects the wrapper in `libfirewall`. Noticeable because `openssh`'s `checkPhase` is now broken which depends on it (didn't notice it earlier because this checkPhase was implemented somewhat recently).

However, `libfirewall` is used by a bunch of other things, so rather than removing the checkPhase again I'll try to find a workaround. Perhaps I'll get to it next week.

Ma27 committed:
glibc: don't hide symbol `__nss_files_fopen`

I realized that `libredirect` doesn't wrap `__nss_files_fopen` anymore
which subsequently broke the `checkPhase` of `pkgs.openssh`[1] which was
implemented in 85c63a515d21c8fa27a561a9b8a1db5b18118e39.

The reason for that is what `libc_hidden_{def,proto}` are used for, to
quote `libc-symbols.h`[2]:

    The following macros are used for PLT bypassing within libc.so
    (and if needed other libraries similarly).

This basically means that these symbols are altered in a way so that
they cannot be wrapped by `libredirect` even though it re-defines
`__nss_files_fopen` while being loaded at first with `LD_PRELOAD`.

The problem was caused by a change in 2.34[3] where the `files`-module
for NSS isn't loaded from `libnss_files.so.2` anymore, but the relevant
symbols are directly obtained from `libc.so.6` since a lot of glibc's
APIs were merged into `libc.so.6` before.

As this would effectively break `libredirect` whenever it tries to wrap
C APIs that interact with NSS (e.g. to use `getpwnam(3)`), I decided to
undo this change for `__nss_files_fopen`.

[1] https://hydra.nixos.org/build/160688523
[2] https://sourceware.org/git/?p=glibc.git;a=blob;f=include/libc-symbols.h;h=d41ecf4384d07f9108698dba1d4e9e6a7a9b975f;hb=ae37d06c7d127817ba43850f0f898b793d42aea7
[3] https://sourceware.org/git/?p=glibc.git;a=commitdiff;h=f9c8b11ed7726b858cd7b7cea0d3d7c5233d78cf;hp=6212bb67f4695962748a5981e1b9fea105af74f6

Ma27 commented:
Rebased onto staging, dropped a few obsolete fixes & fixed the openssh issue in https://github.com/NixOS/nixpkgs/pull/133431/commits/591ba876dbfd68bb6f3995862dc9b960332a8700.

This isn't the only commit I consider worth reviewing, so I'll add the "weird" ones to the OP for reviewers.

Also started a new eval on Hydra.

Ma27 commented:
@vcunat there are plenty of timed-out jobs in https://hydra.nixos.org/jobset/nixpkgs/glibc-2.34, would you mind restarting them? :)

vcunat commented:
I found no better way than to restart first the most common culprits separately, wait until they built (optionally repeat), and finally restart all failures (unfortunately they were too many to pick manually).

Ma27 committed:
libosmscout: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163072881

Ma27 committed:
maude: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163067005

Ma27 committed:
jetbrains.jdk: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163085108

Ma27 committed:
xnee: fix build w/glibc-2.34

ChangeLog: https://hydra.nixos.org/build/163058159

Ma27 committed:
qt512.qtwebengine: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163120450

Ma27 committed:
djmount: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163123869

Ma27 committed:
clingcon: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163641612

Ma27 committed:
mustache-hpp: fix build w/glibc-2.34

Actually, this package doesn't need a build as it copies over the
.hpp-file only, so I simplified it a bit.

Failing Hydra build: https://hydra.nixos.org/build/163643601

Ma27 committed:
recastnavigation: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163634618

Ma27 committed:
spdlog_0: fix build w/glibc-2.34

The patch only appears necessary for spdlog_1, so we skip it for v0
which builds fine without it.

Failing Hydra build: https://hydra.nixos.org/build/163659629

Ma27 committed:
symengine: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163675793

Ma27 commented:
Fixed a few more packages and changed a few older fixes: it bothered me that `catch{,2}` is now in the build-closure of a few packages that don't even run tests (previously it was already an issue that tests were compiled, but never executed). I solved that by either setting `doCheck = true;` - if possible - or by disabling the tests altogether.

Ma27 committed:
zeroc-ice: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163122463

Ma27 committed:
gcc{7,9,10}: apply patches for asan w/glibc-2.34

This should fix a few broken cc-wrapper tests that also check for
libasan[1][2][3]:

    [...]
    checking whether sanitizers are fully functional... ==243==ERROR: AddressSanitizer failed to allocate 0x0 (0) bytes of SetAlternateSignalStack (error code: 22)
    [...]

The underlying issue is that `SIGSTKSZ` isn't a compile-time constant
anymore, but in this case the uninitialized `kAltStackSize` was
initialized early enough to evalute to `0`[4].

The issue is already fixed in gcc11 and there's  GCC 8.5 which also
contains the patch, however the backports to v9 and v10 aren't released
yet, so we have to apply patches on our own here.

For GCC 7.5 I applied the patch from gcc8 as it doesn't seem as if
there's an official upstream backport.

[1] https://hydra.nixos.org/build/163102264
[2] https://hydra.nixos.org/build/163624687
[3] https://hydra.nixos.org/build/163619227
[4] https://gcc.gnu.org/bugzilla/show_bug.cgi?id=100114

SuperSandro2000 commented on nixos/modules/system/boot/stage-1.nix:
```diff
@@ -131,6 +131,26 @@ let
+      # because `kmod` isn't linked against `libpthread.so.0` anymore (since
+      # it was merged into `libc.so.6` since version `2.34`), but still needs
+      # to access it for some reason. This is not an issue in stage-1 itself
+      # because of the  `LD_LIBRARY_PATH`-variable and anytime later because the rpath of
+      # kmod/modprobe points to glibc's `$out/lib` where `libpthread.so.6` exists.
+      # However, this is a problem when the kernel calls `modprobe` inside
+      # the initial ramdisk because it doesn't know about the
+      # `LD_LIbRARY_PATH` and the rpath was nuked.
```
```suggestion
      # because of the `LD_LIBRARY_PATH`-variable and anytime later because the rpath of
      # kmod/modprobe points to glibc's `$out/lib` where `libpthread.so.6` exists.
      # However, this is a problem when the kernel calls `modprobe` inside
      # the initial ramdisk because it doesn't know about the
      # `LD_LIBRARY_PATH` and the rpath was nuked.
```

Ma27 replied:
Done. Commit will be part of the next push.

SuperSandro2000 commented on nixos/modules/system/boot/stage-1-init.sh:
```diff
@@ -227,7 +227,7 @@ done
 ln -s @modulesClosure@/lib/modules /lib/modules
 ln -s @modulesClosure@/lib/firmware /lib/firmware
-echo @extraUtils@/bin/modprobe > /proc/sys/kernel/modprobe
+echo @extraUtils@/bin/modprobe-kernel > /proc/sys/kernel/modprobe
```
```suggestion
# see comment in stage-1.nix for explanation 
echo @extraUtils@/bin/modprobe-kernel > /proc/sys/kernel/modprobe
```

Ma27 replied:
Done. Commit will be part of the nex tpush.

Ma27 committed:
soci: fix build w/glibc-2.34

We don't run the tests and I didn't manage to get them running
trivially, so let's disable them. This not only speeds up
compilation-time, but also fixes the build for glibc 2.34.

Failing Hydra build: https://hydra.nixos.org/build/163952252

Ma27 committed:
sfizz: fix build w/glibc-2.34, actually enable tests

Failing Hydra build: https://hydra.nixos.org/build/163931927

Ma27 committed:
quvi*: mark as broken

Failing Hydra build: https://hydra.nixos.org/build/163943269

Given that it hasn't had a release since 2013[1] and Gentoo/Fedora
already dropped it[2][3]. Also, there are a lot of alternative
stream/video downloaders, so it doesn't seem a big deal.

[1] http://quvi.sourceforge.net/news/
[2] https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=80024f2367f394b46326064823d0c39f373e2820
[3] https://src.fedoraproject.org/rpms/quvi/c/bf189503ae6c15d2731679e292c0a90ce1454e62?branch=rawhide

Ma27 committed:
zeroc-ice-36: remove, unmaintained

Ma27 committed:
jumanpp: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163976320

Ma27 committed:
cpp-hocon: fix build w/glibc-2.34

Simply disable tests as it tries to pull in an old `catch2` from
`leatherman`.

Failing Hydra build: https://hydra.nixos.org/build/163982252

Ma27 committed:
cataclysm-dda: fix build w/glibc-2.34

* Enable parallel building to speed up the build-time.
* Disable tests as vendored catch2 doesn't compile against glibc 2.34
  and I couldn't get the tests to run in the sandbox.

Failing Hydra build: https://hydra.nixos.org/build/163952121

Ma27 committed:
securefs: fix build w/glibc-2.34

Tests don't work in the sandbox and vendored catch2 doesn't compile, so
we disable the tests.

Failing Hydra build: https://hydra.nixos.org/build/163981825

Ma27 committed:
sfxr-qt: fix build w/glibc-2.34

Failing Hydra build: https://hydra.nixos.org/build/163999266

Ma27 commented:
I'd consider this in a reviewable state now, although we should await at least one more Hydra eval and I think it makes sense if I test it on my laptop as this was a more complex update and I'd like to see if everything remains smooth :) 

However, I'd mark it as "ready for review" now.

jonringer commented:
To help move this along, I created a hydra job https://hydra.jonringer.us/jobset/nixpkgs/glibc-2.34

vcunat commented:
OK, I guess I'll cancel the one on hydra.nixos.org.  I assume yours will substitute from cache.nixos.org but it doesn't work the other way around.

jonringer commented:
> OK, I guess I'll cancel the one on hydra.nixos.org. I assume yours will substitute from cache.nixos.org but it doesn't work the other way around.

Correct.

Trying to get another staging-next going

jonringer commented:
On another note. glibc 2.35 was released. Should we leapfrog?

jonringer commented:
> If 2.34 is (almost) ready, I'd go with that first.

*re-looks at the 60+ regression fixes*

Yea, we can save that for another PR. This has already been going on for several months.

Ma27 commented:
> To help move this along, I created a hydra job https://hydra.jonringer.us/jobset/nixpkgs/glibc-2.34

Perhaps I missed something, but why don't we use hydra.nixos.org anymore?

> re-looks at the 60+ regression fixes

Ack. This update was especially complex. That's also why I'm planning to rebuil my laptop against this branch for testing purposes :)

jonringer commented:
> Perhaps I missed something, but why don't we use hydra.nixos.org anymore?

As of late, official hydra seems to be getting more work than it can handle. I would like to reserve it's capacity for release channels and staging workflows.

PR specific jobsets which are mostly just for regression finding can be done anywhere. Having the builds pushed to `cache.nixos.org` is a nice plus, but only really relevant during the lifetime of a PR.

committers/contributors are free to add themselves for ssh access: https://github.com/jonringer/server-configuration#ringers-server

jonringer commented:
once rebased, I think this is good to go. The hydra jobset looks like staging-next before we started doing stabilizaiton.

Used nix-review-tool to see if there was any major breakages

Ma27 commented:
Agreed, but

* I'd still like to throw this against e.g. my workstation first
* It could use some reviews (*especially* the four commits I referenced in my OP :))

jonringer commented:
sounds good

vcunat commented on pkgs/development/libraries/leatherman/default.nix:
```diff
@@ -11,6 +11,10 @@ stdenv.mkDerivation rec {
   };
 
+  preBuild = ''
+    cp ${catch}/include/catch/catch.hpp src/Catch-1.10.0/single_include/catch.hpp
```
Also, nix store and temporary directory may commonly be separate filesystems (e.g. `tmpOnTmpfs`).

vcunat commented:
Maintainer pings, just to be sure:
- @abbradar: `zeroc-ice-36` removed
- @Hodapp87, @kovirobi: `libfive` marked as broken

trofi commented on pkgs/development/libraries/glibc/default.nix:
```diff
@@ -119,6 +119,16 @@ callPackage ./common.nix { inherit stdenv; } {
+      # because it's not actually needed anymore to link against `pthreads` since
+      # it's now part of `libc.so.6` itself, but the gcc build breaks if
+      # this doesn't work.
+      ln -sf $out/lib/libpthread.so.0 $out/lib/libpthread.so
```
Upstream recommends installing empty `.a` files like `libpthread.a` for future bianries to avoid dynamic linking against these libraries.

Ma27 replied:
Done, change will be part of the next push.

However, this doesn't fix the initrd issue on its own: `${pkgs.kmod}` doesn't have `libpthread` in its `DT_NEEDED` and thus doesn't seem to be linked against it anymore. Also, due to no other references, `libpthread.a` isn't part of `$out/lib` from `extra-utils`, so the hack is still needed IMO.

trofi commented on pkgs/development/tools/misc/gdb/default.nix:
```diff
@@ -92,6 +92,7 @@ stdenv.mkDerivation rec {
     "--disable-shared" "--enable-static"
     "--with-system-zlib"
     "--with-system-readline"
+    "--disable-sim"
```
Perhaps just updating `gdb` is less drastic than disabling `sim` entirely: https://github.com/NixOS/nixpkgs/pull/155233
(and while at it apply https://sourceware.org/pipermail/binutils/2022-February/119742.html)

Ma27 replied:
In fact I can now compile `gdb` on this branch without this patch.... I'll push the change with the next batch of changes :)

trofi commented on pkgs/stdenv/linux/default.nix:
```diff
@@ -254,6 +254,19 @@ in
         '';
       });
 
+      gcc-unwrapped = self.stdenvNoCC.mkDerivation {
+        inherit (prevStage.gcc-unwrapped) name;
+        dontUnpack = true;
+        dontBuild = true;
+        installPhase = ''
+          mkdir -p $out
+          cp -ar ${prevStage.gcc-unwrapped}/* $out/
+          chmod -R a+w $out/
+          cp -a ${getLibc self}/lib/* $out/lib/
+          chmod -R a-w $out/
+        '';
+      };
+
```
I encountered something similar in https://github.com/NixOS/nixpkgs/issues/158042 (when tried glibc-2.35 locally). I think the bug is in order or -L/-B options. I managed to get away with it by fixing cflags order: https://github.com/NixOS/nixpkgs/pull/158047

Ma27 replied:
The change seems to be a fix indeed, thanks!

Ma27 commented:
> Perhaps just updating gdb is less drastic than disabling sim entirely: #155233

At the time I encountered this, there was IIRC no patch or release for this. But if there's one now, it's of course the better option :)

Ma27 commented:
Now that we've rebased I'll await a new Hydra iteration on @jonringer's Hydra to get a binary cache with all build products. Then, I'd take a look at the suggestions from @trofi %)

jonringer commented:
@GrahamcOfBorg eval

Ma27 commented:
@jonringer would you mind restarting hte aborted jobs in https://hydra.jonringer.us/eval/551 ? :)

Ma27 committed:
gdb: remove `--disable-sim` fix

Ma27 committed:
texinfo: review fixes

Ma27 commented on pkgs/development/tools/misc/texinfo/common.nix:
```diff
@@ -26,7 +26,7 @@ stdenv.mkDerivation {
   };
 
-  patches = optional crossBuildTools ./cross-tools-flags.patch;
+  patches = patches ++ (optional crossBuildTools ./cross-tools-flags.patch);
```
Fixed locally, change will be part of my next push.

Ma27 committed:
glibc: add empty libpthread.a

See https://github.com/NixOS/nixpkgs/pull/133431#discussion_r805338099

Ma27 committed:
nixos/stage-1: typo fixes

Ma27 committed:
stdenv: revert gcc hack

See https://github.com/NixOS/nixpkgs/pull/133431#discussion_r805339569

Ma27 committed:
Revert "jetbrains.jdk: fix build w/glibc-2.34"

This reverts commit 4427e547ca418bac7cc2a6b4c5d144a4a2fac126.

Patch is not needed anymore.

jonringer commented:
> @jonringer would you mind restarting hte aborted jobs in https://hydra.jonringer.us/eval/551 ? :)

Sorry I just say this now. Restarted

Ma27 commented:
TIL to build my entire workstation's NixOS from source, I need a c6a.32xlarge from AWS and ~90GB of free disk space. In other words, I'm currently writing this comment from a NixOS using glibc 2.34.

Since that's also working perfectly fine, I'd say that this PR has reached a reasonably stable state. @jonringer I guess it'd make sense to cancel the jobset again and start a new eval with the current state.

As soon as everything else is sorted out, I'd be fine with merging :)

Ma27 committed:
cataclysm-dda: fix eval

No idea why this breaks now, but let's see if that's sufficient to make
ofborg happy.

jonringer commented:
I think we can get the remaining regressions in staging-next. The next round should probably be dedicated to glibc, then the round after for systemd 250.

jonringer commented:
cc @NixOS/nixos-release-managers

vcunat commented:
So the recent hydra.nixos.org evaluation of this should be cancelled?  I can't see why it's building on both farms at once.

Ma27 commented:
I think that was on oversight of mine, cancelled on hydra.nixos.org, sorry!

jonringer reviewed (approved):
LGTM

Result of nixpkgs-review pr 133431 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>2 packages built:</summary>  <ul>    <li>python3Packages.numpy</li>    <li>ripgrep</li>  </ul></details>

jonringer commented:
```
https://github.com/NixOS/nixpkgs/pull/133431

2 packages built:
python3Packages.numpy ripgrep
```

trofi commented:
I think `stdenv` changes were not completely rolled back. Noticed as a bootstrap failure on glibc-2.35. Proposed the fix as https://github.com/NixOS/nixpkgs/pull/162197

Ma27 commented:
also, does that mean you're working on glibc-2.35? :) 

(in that case the next mass-rebuild project for me would be a GCC major bump %))

trofi commented:
I run `glibc-2.35` only locally to have a fresh upstream release. I don't think i'll be able to get all of `nixpkgs` to `2.35`. I can only provide one-off fixes for packages I use.
